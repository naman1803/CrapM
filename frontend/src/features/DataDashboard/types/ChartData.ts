// used for SeverityBarChart
export interface SeverityData {
    facilityOwner: string;
    severityLevel: string;
    severityScore: number;
}

// used for SeverityOverTimeChart
export interface SeverityOverTimeData extends SeverityData {
    date: string;
}

// one instance of ranking
export interface FacilityExceedanceData {
    facilityOwner: string;
    noOfExceedances: number;
}
