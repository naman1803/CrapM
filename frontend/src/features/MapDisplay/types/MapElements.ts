import { LatLngExpression } from 'leaflet';

export enum MapElementType {
  Marker = 'marker',
  Circle = 'circle',
}

// Represent a point on the map
interface MapElementBase {
  id: number;
  dataType: string;
  isSelectable: boolean;
}

interface MarkerElement extends MapElementBase {
  elementType: MapElementType.Marker;
  dataPoint: GeoPoint;
}

interface CircleElement extends MapElementBase {
  elementType: MapElementType.Circle;
  dataPoint: GeoCircle;
}

export interface GeoPoint {
  coordinates: LatLngExpression;
}

interface GeoCircle extends GeoPoint {
  radius: number;
  colour: string;
}

export interface LakeGeoPoint extends GeoPoint {
  id: number;
  name: string;
  additional_notes?: string;
}

export interface SewageGeoPoint extends GeoCircle {
  id: number;
  facility_owner: string;
}

export type MapElement = MarkerElement | CircleElement;
