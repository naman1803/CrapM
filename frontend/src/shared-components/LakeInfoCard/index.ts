export { default as LakeInfoCard } from './LakeInfoCard';
export type { LakeInfoCardProps } from './LakeInfoCard';
