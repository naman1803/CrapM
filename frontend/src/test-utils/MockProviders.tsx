import { ThemeProvider } from '@mui/material';
import { MemoryRouter } from 'react-router-dom';
import { themeOptions } from '@app/theme';
import React from 'react';

// Wraps all providers for the app for testing
export const TestProviders = ({
  children,
  initialRoute = '/',
}: {
  children?: React.ReactNode;
  initialRoute?: string;
}) => {
  return (
    <ThemeProvider theme={themeOptions}>
      <MemoryRouter initialEntries={[initialRoute]}>{children}</MemoryRouter>
    </ThemeProvider>
  );
};
