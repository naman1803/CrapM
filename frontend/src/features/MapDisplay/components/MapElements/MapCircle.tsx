import { Circle } from 'react-leaflet';
import { MapElement, SewageGeoPoint } from '../../types/MapElements';
import React, { useState, useEffect, useRef } from 'react';
import { LatLngTuple } from 'leaflet';

interface MapElementCircleProps {
  element: MapElement;
  onElementSelect: (elementSelected: MapElement) => void;
  hoveredPoint: MapElement | null;
  selectedPoint: MapElement | null;
}
const MapElementCircle: React.FC<MapElementCircleProps> = ({
  element,
  onElementSelect,
  hoveredPoint,
  selectedPoint,
}) => {
  const datapoint = element.dataPoint as SewageGeoPoint;
  const [color, setColor] = useState('gray'); // Default color
  const baseColor = useRef('gray');

  // Change color based on hover
  useEffect(() => {
    if (
      selectedPoint &&
      selectedPoint.dataPoint.coordinates === datapoint.coordinates
    ) {
      setColor('red'); // Selected color
      baseColor.current = 'red';
    } else if (
      hoveredPoint &&
      hoveredPoint.dataPoint.coordinates === datapoint.coordinates
    ) {
      setColor('blue'); // Hover color
      baseColor.current = 'blue';
    } else {
      setColor(datapoint.colour); // Default color
      baseColor.current = datapoint.colour;
    }
  }, [selectedPoint, hoveredPoint, datapoint.coordinates, datapoint.colour]);

  return (
    <Circle
      radius={isNaN(datapoint?.radius) ? 20 : datapoint.radius * 20}
      color={datapoint.colour}
      center={datapoint.coordinates}
      data-testid={`Circle-${datapoint.coordinates as LatLngTuple}`}
      pathOptions={{ color }}
      eventHandlers={{
        ...(element.isSelectable
          ? {
              click: () => onElementSelect(element),
              mouseover: () => setColor('blue'), // Hover color
              mouseout: () => setColor(baseColor.current), // Revert to base color
            }
          : {}),
      }}
    ></Circle>
  );
};

export default MapElementCircle;
