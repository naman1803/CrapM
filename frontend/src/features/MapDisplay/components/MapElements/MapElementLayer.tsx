import { MapElementType, MapElement } from '../../types/MapElements';
import MapElementMarker from './MapMarker';
import MapElementCircle from './MapCircle';

export interface MapElementLayerProps {
  elements: MapElement[];
  onElementClick: (element: MapElement) => void;
  hoveredPoint: MapElement | null;
  selectedPoint: MapElement | null;
  zoom: number;
}

const MapElementLayer: React.FC<MapElementLayerProps> = ({
  elements,
  onElementClick,
  hoveredPoint,
  selectedPoint,
  zoom,
}) => {
  return (
    <>
      {elements.map((element) => {
        switch (element.elementType) {
          case MapElementType.Marker:
            return (
              <MapElementMarker
                key={`${element.id}-${element.elementType}`}
                element={element}
                onElementSelect={onElementClick}
                hoveredPoint={hoveredPoint}
                selectedPoint={selectedPoint}
                zoom={zoom}
              />
            );
          case MapElementType.Circle:
            return (
              <MapElementCircle
                key={`${element.id}-${element.elementType}`}
                element={element}
                onElementSelect={onElementClick}
                hoveredPoint={hoveredPoint}
                selectedPoint={selectedPoint}
              />
            );
          default:
            return <></>;
        }
      })}
    </>
  );
};
export default MapElementLayer;
