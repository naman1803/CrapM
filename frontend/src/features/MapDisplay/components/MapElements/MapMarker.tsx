import { Marker } from 'react-leaflet';
import { MapElement } from '../../types/MapElements';
import React, { useEffect, useState, useRef } from 'react';
import L from 'leaflet';

interface MapElementMarkerProps {
  element: MapElement;
  onElementSelect?: (elementSelected: MapElement) => void;
  hoveredPoint: MapElement | null;
  selectedPoint: MapElement | null;
  zoom: number;
}

const MapElementMarker: React.FC<MapElementMarkerProps> = ({
  element,
  onElementSelect,
  hoveredPoint,
  selectedPoint,
  zoom
}) => {
  const [color, setColor] = useState('gray');  // Default color
  const baseColor = useRef('gray');

  // Change color based on hover
  useEffect(() => {
    if (selectedPoint && selectedPoint.dataPoint.coordinates === element.dataPoint.coordinates) {
      setColor('red');  // Selected color
      baseColor.current = 'red';
    } else if (hoveredPoint && hoveredPoint.dataPoint.coordinates === element.dataPoint.coordinates) {
      setColor('blue');  // Hover color
      baseColor.current = 'blue';
    } else {
      setColor('gray');  // Default color
      baseColor.current = 'gray';
    }
  }, [selectedPoint, hoveredPoint, element.dataPoint.coordinates]);
  
  const MIN_SIZE = 7; 

  const markerSize = Math.max(MIN_SIZE, zoom*1.5);
  // Custom HTML-based icon with dynamic color
  const markerHtmlStyles = `
    background-color: ${color};
    width: ${markerSize}px;
    height: ${markerSize}px;
    display: block;
    left: -${markerSize / 2}px;  /* Center horizontally */
    top: -${markerSize / 2}px;   /* Center vertically */
    position: relative;
    border-radius: 3rem 3rem 0;
    transform: rotate(45deg);
    border: 1px solid #FFFFFF;
  `;

  const icon = L.divIcon({
    className: 'lake-marker',
    iconAnchor: [0, 10],
    popupAnchor: [0, -36],
    html: `<span style="${markerHtmlStyles}" />`,
  });

  return (
    <Marker
      position={element.dataPoint.coordinates}
      icon={icon}
      eventHandlers={{
        ...(element.isSelectable
          ? {
              click: () => onElementSelect && onElementSelect(element),
              mouseover: () => setColor('blue'), // Hover color
              mouseout: () => setColor(baseColor.current) // Revert to base color
            }
          : {})
      }}
    />

  );
};

export default MapElementMarker;