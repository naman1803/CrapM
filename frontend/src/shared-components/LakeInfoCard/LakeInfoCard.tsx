import { InfoCard } from '../InfoCard';
import { LakeGeoPoint, MapElement } from '@/features/MapDisplay/types/MapElements';
import { LatLngTuple } from 'leaflet';
export interface LakeInfoCardProps {
  lake: MapElement;
  handleClose: () => void;
}

/** An info card which displays information about a given lake*/
const LakeInfoCard: React.FC<LakeInfoCardProps> = ({ lake, handleClose }) => {
  const point: LakeGeoPoint = lake.dataPoint as LakeGeoPoint;
  const coords: LatLngTuple = point.coordinates as LatLngTuple;
  const infoItems = [
    `Latitude: ${coords[0]}`,
    `Longitude: ${coords[1]}`,
    ...(point.additional_notes
      ? [`Misc Information: ${point.additional_notes}`]
      : []),
  ];

  return (
    <InfoCard
      title={point.name}
      infoItems={infoItems}
      handleClose={handleClose}
    >
      <></>
    </InfoCard>
  );
};

export default LakeInfoCard;
