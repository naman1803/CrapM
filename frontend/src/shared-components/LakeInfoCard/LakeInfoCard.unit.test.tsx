import { render } from '@testing-library/react';
import { TestProviders } from '@/test-utils/MockProviders';
import LakeInfoCard, { LakeInfoCardProps } from './LakeInfoCard';
import { LakeGeoPoint, MapElement, MapElementType } from '@/features/MapDisplay/types/MapElements';

const lakePoint = {
  id: 1,
  name: 'Mynor Lake',
  coordinates: [123, 456],
  additional_notes: 'Test Notes',
} as LakeGeoPoint;
const mockProps: LakeInfoCardProps = {
  lake: {
    id: 1,
    dataType: 'lake',
    isSelectable: true,
    elementType: MapElementType.Marker,
    dataPoint: lakePoint
  },
  handleClose: jest.fn(),
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe('LakeInfoCard', () => {
  it('renders correctly', () => {
    const { asFragment } = renderLakeInfoCard(mockProps);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders elements', () => {
    const { getByText, getByRole } = renderLakeInfoCard(mockProps);
    expect(getByText('Mynor Lake')).toBeInTheDocument();
    expect(getByText('Latitude: 123')).toBeInTheDocument();
    expect(getByText('Longitude: 456')).toBeInTheDocument();
    expect(getByText('Misc Information: Test Notes')).toBeInTheDocument();
    expect(getByRole('button')).toBeInTheDocument();
  });
  it('calls handleClose when close button is clicked', () => {
    const { getByRole } = renderLakeInfoCard(mockProps);
    getByRole('button').click();
    expect(mockProps.handleClose).toHaveBeenCalled();
  });
  it('does not show misc information if there are no additional notes', () => {
    const newLakePoint = { ...lakePoint, additional_notes: undefined };
    const { queryByText } = renderLakeInfoCard({
      ...mockProps,
      lake: { 
        ...mockProps.lake, 
        dataPoint: newLakePoint
      } as MapElement,
    });
    expect(queryByText('Misc Information')).not.toBeInTheDocument();
  });
});

const renderLakeInfoCard = (props: LakeInfoCardProps) => {
  return render(
    <TestProviders>
      <LakeInfoCard {...props} />
    </TestProviders>,
  );
};
