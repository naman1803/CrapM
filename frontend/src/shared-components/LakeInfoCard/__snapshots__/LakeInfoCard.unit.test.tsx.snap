// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LakeInfoCard renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-29of7v"
    data-testid="infocard-body"
  >
    <div
      class="MuiBox-root css-16yr8je"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 css-dvb9hv-MuiTypography-root"
      >
        Mynor Lake
      </h4>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1a49dbd-MuiButtonBase-root-MuiIconButton-root"
        data-testid="infocard-exit"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="CloseIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </button>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-x148m1-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-1y4jlh3"
    >
      <div
        class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-spacing-xs-2 css-ubb9a5-MuiGrid2-root"
      >
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-6 css-ticvdu-MuiGrid2-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-tckkrt-MuiTypography-root"
          >
            Latitude: 123
          </p>
        </div>
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-6 css-ticvdu-MuiGrid2-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-tckkrt-MuiTypography-root"
          >
            Longitude: 456
          </p>
        </div>
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-6 css-ticvdu-MuiGrid2-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-tckkrt-MuiTypography-root"
          >
            Misc Information: Test Notes
          </p>
        </div>
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-1j0u08r-MuiGrid2-root"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
